File: server/src/main/java/sa/elm/mashrook/pledges/PledgeService.java
package sa.elm.mashrook.pledges;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import sa.elm.mashrook.campaigns.domain.CampaignEntity;
import sa.elm.mashrook.campaigns.domain.CampaignStatus;
import sa.elm.mashrook.exceptions.InvalidCampaignStateException;
import sa.elm.mashrook.exceptions.InvalidPledgeStateException;
import sa.elm.mashrook.exceptions.OrganizationPendingVerificationException;
import sa.elm.mashrook.exceptions.PledgeAccessDeniedException;
import sa.elm.mashrook.exceptions.PledgeAlreadyExistsException;
import sa.elm.mashrook.exceptions.PledgeNotFoundException;
import sa.elm.mashrook.organizations.domain.OrganizationEntity;
import sa.elm.mashrook.organizations.domain.OrganizationStatus;
import sa.elm.mashrook.pledges.domain.PledgeEntity;
import sa.elm.mashrook.pledges.domain.PledgeStatus;
import sa.elm.mashrook.pledges.dto.PledgeCreateRequest;
import sa.elm.mashrook.pledges.dto.PledgeListResponse;
import sa.elm.mashrook.pledges.dto.PledgeResponse;
import sa.elm.mashrook.pledges.dto.PledgeUpdateRequest;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Set;
import java.util.UUID;

@Slf4j
@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class PledgeService {

    private final PledgeRepository pledgeRepository;

    @Transactional
    public PledgeResponse createPledge(CampaignEntity campaign,
                                       OrganizationEntity buyerOrg,
                                       PledgeCreateRequest request) {
        validateOrganizationIsActive(buyerOrg);
        validateCampaignAcceptsPledges(campaign);

        // Check if a pledge already exists for this campaign and organization
        var existingPledge = pledgeRepository.findByCampaignIdAndOrganizationId(
                campaign.getId(), buyerOrg.getId());

        if (existingPledge.isPresent()) {
            PledgeEntity pledge = existingPledge.get();

            // If the existing pledge is withdrawn, reactivate it
            if (pledge.getStatus() == PledgeStatus.WITHDRAWN) {
                log.info("Reactivating withdrawn pledge {} for campaign {} by organization {}",
                        pledge.getId(), campaign.getId(), buyerOrg.getId());
                pledge.setStatus(PledgeStatus.PENDING);
                pledge.setQuantity(request.quantity());
                pledge.setCommittedAt(null); // Clear any previous commit timestamp
                PledgeEntity saved = pledgeRepository.save(pledge);
                return PledgeResponse.from(saved);
            }

            // If pledge exists and is not withdrawn, throw duplicate error
            throw new PledgeAlreadyExistsException(
                    String.format("A pledge already exists for campaign %s by buyer %s",
                            campaign.getId(), buyerOrg.getId()));
        }

        // No existing pledge, create a new one
        PledgeEntity pledge = new PledgeEntity();
        pledge.setCampaign(campaign);
        pledge.setOrganization(buyerOrg);
        pledge.setQuantity(request.quantity());
        pledge.setStatus(PledgeStatus.PENDING);

        PledgeEntity saved = pledgeRepository.save(pledge);

        return PledgeResponse.from(saved);
    }

    @Transactional
    public PledgeResponse updatePledge(UUID pledgeId, UUID buyerOrgId, PledgeUpdateRequest request) {
        PledgeEntity pledge = findPledgeOrThrow(pledgeId);
        validatePledgeOwnership(pledge, buyerOrgId);

        // Pledges can only be updated during ACTIVE phase
        validateCampaignIsActive(pledge.getCampaign());

        pledge.setQuantity(request.quantity());
        PledgeEntity saved = pledgeRepository.save(pledge);
        return PledgeResponse.from(saved);
    }

    @Transactional
    public void cancelPledge(UUID pledgeId, UUID buyerOrgId) {
        PledgeEntity pledge = findPledgeOrThrow(pledgeId);
        validatePledgeOwnership(pledge, buyerOrgId);

        if (PledgeStatus.WITHDRAWN.equals(pledge.getStatus())) {
            log.debug("pledge: {} is already withdrawn", pledge.getId());
            return;
        }

        // Pledges can only be cancelled during ACTIVE phase
        validateCampaignIsActive(pledge.getCampaign());

        pledge.setStatus(PledgeStatus.WITHDRAWN);
        pledgeRepository.save(pledge);
    }

    @Transactional
    public PledgeResponse commitPledge(UUID pledgeId, UUID buyerOrgId) {
        PledgeEntity pledge = findPledgeOrThrow(pledgeId);
        validatePledgeOwnership(pledge, buyerOrgId);
        validateCampaignInGracePeriod(pledge.getCampaign());
        validatePledgeIsPending(pledge);

        pledge.setStatus(PledgeStatus.COMMITTED);
        pledge.setCommittedAt(LocalDateTime.now());

        PledgeEntity saved = pledgeRepository.save(pledge);
        return PledgeResponse.from(saved);
    }

    private void validateCampaignInGracePeriod(CampaignEntity campaign) {
        if (campaign.getStatus() != CampaignStatus.GRACE_PERIOD) {
            throw new InvalidCampaignStateException(
                    String.format("Pledge commitment is only allowed during GRACE_PERIOD. Campaign %s is in %s status",
                            campaign.getId(), campaign.getStatus()));
        }
    }

    private void validatePledgeIsPending(PledgeEntity pledge) {
        if (pledge.getStatus() != PledgeStatus.PENDING) {
            throw new InvalidPledgeStateException(
                    String.format("Only PENDING pledges can be committed. Pledge %s is in %s status",
                            pledge.getId(), pledge.getStatus()));
        }
    }

    public PledgeListResponse getBuyerPledges(UUID buyerOrgId, String status, Pageable pageable) {
        Page<PledgeEntity> pledges;

        if (status != null) {
            // Filter by specific status when provided
            pledges = pledgeRepository.findAllByOrganizationIdAndStatus(buyerOrgId, PledgeStatus.getStatus(status), pageable);
        } else {
            // By default, return all non-withdrawn pledges (PENDING + COMMITTED)
            pledges = pledgeRepository.findAllByOrganizationIdAndStatusNot(buyerOrgId, PledgeStatus.WITHDRAWN, pageable);
        }

        Page<PledgeResponse> responsePage = pledges.map(PledgeResponse::from);
        return PledgeListResponse.from(responsePage);
    }

    public PledgeListResponse getCampaignPledges(CampaignEntity campaign, Pageable pageable) {
        Page<PledgeEntity> pledges = pledgeRepository.findAllByCampaignId(campaign.getId(), pageable);
        Page<PledgeResponse> responsePage = pledges.map(PledgeResponse::from);
        return PledgeListResponse.from(responsePage);
    }

    private PledgeEntity findPledgeOrThrow(UUID pledgeId) {
        return pledgeRepository.findById(pledgeId)
                .orElseThrow(() -> new PledgeNotFoundException(
                        String.format("Pledge with id %s not found", pledgeId)));
    }

    private void validateOrganizationIsActive(OrganizationEntity organization) {
        if (organization.getStatus() != OrganizationStatus.ACTIVE) {
            throw new OrganizationPendingVerificationException("Organization is not active");
        }
    }

    private void validateCampaignAcceptsPledges(CampaignEntity campaign) {
        Set<CampaignStatus> pledgeAllowedStatuses = Set.of(CampaignStatus.ACTIVE, CampaignStatus.GRACE_PERIOD);
        if (!pledgeAllowedStatuses.contains(campaign.getStatus())) {
            throw new InvalidCampaignStateException(
                    String.format("Campaign %s is not accepting pledges. Current status: %s",
                            campaign.getId(), campaign.getStatus()));
        }
    }

    private void validateCampaignIsActive(CampaignEntity campaign) {
        if (campaign.getStatus() != CampaignStatus.ACTIVE) {
            throw new InvalidCampaignStateException(
                    String.format("Pledges can only be modified during ACTIVE phase. Campaign %s is in %s status",
                            campaign.getId(), campaign.getStatus()));
        }
    }

    private void validatePledgeOwnership(PledgeEntity pledge, UUID buyerOrgId) {
        if (!pledge.getOrganization().getId().equals(buyerOrgId)) {
            throw new PledgeAccessDeniedException("You do not have permission to modify this pledge");
        }
    }

    public List<PledgeEntity> findAllByCampaignIdAndStatus(UUID campaignId, PledgeStatus pledgeStatus) {
        return pledgeRepository.findAllByCampaignIdAndStatus(campaignId, pledgeStatus);
    }

    public int calculateTotalCommitedPledges(UUID campaignId) {
        return findAllByCampaignIdAndStatus(campaignId, PledgeStatus.COMMITTED)
                .stream()
                .mapToInt(PledgeEntity::getQuantity)
                .sum();
    }

    /**
     * Calculates the total pledged quantity for a campaign, including both PENDING and COMMITTED pledges.
     * This is used for displaying current pricing to buyers.
     *
     * @param campaignId the ID of the campaign
     * @return the total quantity of all non-withdrawn pledges
     */
    public int calculateTotalActivePledges(UUID campaignId) {
        return pledgeRepository.findAllByCampaignId(campaignId).stream()
                .filter(pledge -> pledge.getStatus() != PledgeStatus.WITHDRAWN)
                .mapToInt(PledgeEntity::getQuantity)
                .sum();
    }

    /**
     * Withdraws all PENDING pledges for a given campaign.
     *
     * <p>This method is called when a campaign transitions from GRACE_PERIOD
     * to LOCKED or CANCELLED. All pledges that were not explicitly committed
     * during the grace period are automatically withdrawn.
     *
     * @param campaignId the ID of the campaign
     */
    @Transactional
    public void withdrawAllPendingPledges(UUID campaignId) {
        List<PledgeEntity> pendingPledges = findAllByCampaignIdAndStatus(campaignId, PledgeStatus.PENDING);

        pendingPledges.forEach(pledge -> pledge.setStatus(PledgeStatus.WITHDRAWN));

        pledgeRepository.saveAll(pendingPledges);
    }
}

File: client/app/services/pledge.service.ts
/**
 * Pledge Service
 *
 * Provides pledge management functionality for buyers.
 * Communicates with the Spring Boot backend pledge endpoints.
 */

import { apiClient } from "@/lib/api-client";
import type { PledgeStatus } from "@/types/campaign";

/**
 * Request payload for creating a pledge
 */
export interface PledgeCreateRequest {
	quantity: number;
}

/**
 * Request payload for updating a pledge
 */
export interface PledgeUpdateRequest {
	quantity: number;
}

/**
 * Response from pledge API endpoints
 */
export interface PledgeResponse {
	id: string;
	campaignId: string;
	buyerOrgId: string;
	quantity: number;
	status: PledgeStatus;
	committedAt: string | null;
	createdAt: string;
	updatedAt: string;
}

/**
 * Paginated list response for pledges
 */
export interface PledgeListResponse {
	content: PledgeResponse[];
	page: number;
	size: number;
	totalElements: number;
	totalPages: number;
}

/**
 * Options for getting buyer's pledges
 */
export interface GetBuyerPledgesOptions {
	status?: PledgeStatus;
	page?: number;
	size?: number;
}

/**
 * Options for getting campaign pledges
 */
export interface GetCampaignPledgesOptions {
	page?: number;
	size?: number;
}

/**
 * Build query string for buyer pledges
 */
function buildBuyerPledgesQueryString(options: GetBuyerPledgesOptions): string {
	const params: string[] = [];

	if (options.status) {
		params.push(`status=${encodeURIComponent(options.status)}`);
	}

	if (options.page !== undefined) {
		params.push(`page=${options.page}`);
	}

	if (options.size !== undefined) {
		params.push(`size=${options.size}`);
	}

	return params.length > 0 ? `?${params.join("&")}` : "";
}

/**
 * Build query string for campaign pledges
 */
function buildCampaignPledgesQueryString(options: GetCampaignPledgesOptions): string {
	const params: string[] = [];

	if (options.page !== undefined) {
		params.push(`page=${options.page}`);
	}

	if (options.size !== undefined) {
		params.push(`size=${options.size}`);
	}

	return params.length > 0 ? `?${params.join("&")}` : "";
}

export const pledgeService = {
	async createPledge(
		campaignId: string,
		data: PledgeCreateRequest
	): Promise<PledgeResponse> {
		return apiClient.post<PledgeResponse>(
			`/v1/pledges/campaigns/${campaignId}`,
			data
		);
	},

	/**
	 * Update an existing pledge
	 *
	 * @param campaignId - Campaign ID
	 * @param pledgeId - Pledge ID
	 * @param data - Updated pledge data with quantity
	 * @returns Updated pledge response
	 * @throws Error if pledge not found or not authorized
	 */
	async updatePledge(
		campaignId: string,
		pledgeId: string,
		data: PledgeUpdateRequest
	): Promise<PledgeResponse> {
		return apiClient.put<PledgeResponse>(
			`/v1/campaigns/${campaignId}/pledges/${pledgeId}`,
			data
		);
	},

	/**
	 * Cancel a pledge
	 *
	 * @param campaignId - Campaign ID
	 * @param pledgeId - Pledge ID
	 * @throws Error if pledge not found or not authorized
	 */
	async cancelPledge(campaignId: string, pledgeId: string): Promise<void> {
		return apiClient.delete(`/v1/campaigns/${campaignId}/pledges/${pledgeId}`);
	},

	/**
	 * Get buyer's pledges with optional filters and pagination
	 *
	 * @param options - Optional filters for status and pagination
	 * @returns Paginated list of pledges
	 */
	async getBuyerPledges(
		options: GetBuyerPledgesOptions = {}
	): Promise<PledgeListResponse> {
		const queryString = buildBuyerPledgesQueryString(options);
		return apiClient.get<PledgeListResponse>(`/v1/pledges${queryString}`);
	},

	/**
	 * Get pledges for a specific campaign
	 *
	 * @param campaignId - Campaign ID
	 * @param options - Optional pagination options
	 * @returns Paginated list of pledges for the campaign
	 * @throws Error if campaign not found
	 */
	async getCampaignPledges(
		campaignId: string,
		options: GetCampaignPledgesOptions = {}
	): Promise<PledgeListResponse> {
		const queryString = buildCampaignPledgesQueryString(options);
		return apiClient.get<PledgeListResponse>(
			`/v1/pledges/campaigns/${campaignId}${queryString}`
		);
	},

	/**
	 * Commit a pledge during the grace period
	 *
	 * @param pledgeId - Pledge ID
	 * @returns Committed pledge response with updated status
	 * @throws Error if pledge not found, not authorized, or campaign not in grace period
	 */
	async commitPledge(pledgeId: string): Promise<PledgeResponse> {
		return apiClient.post<PledgeResponse>(`/v1/pledges/${pledgeId}/commit`);
	},
};
